### PR TITLE
Start/restart/stop priority feature implementation

### PR DESCRIPTION
**Pull request by livankrekh**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1234, #5678
| License       | MIT
| Doc PR        | https://github.com/pm2-hive/pm2-hive.github.io/pulls

### CODE REVIEW

**Comment by vmarchaud** [comment]
Is this PR still alive ? Can you explain a little bit more about what you want to achieve ?

**Comment by livankrekh** [comment]
The main idea of this PR and feature: starting and stoping processes by ecosystem order (priority and stop_priority properties)
That realized in the next way:
1) List of processes (appConf, apps_to_start and others) sorting by this priority (by checkPriority function). If processes have same priority - these processes collecting in array, that included in process list (for example: process1.priority: 2, process2.priority: 1, process3.priority: 20, process4.priority: 2. Result will be: [process2, [process1, process4], process3] )
2. That processes list will be handled by async.eachSeries (all processes will be executed 
consistently). If async.eachSeries get a array of processes (when processes have same priority), in this case each process of array will be executed parallel
3. **eachSeries starts next process, if previous process called callback. Callback calling, when child process send signal 'pm2:started' or 'ready', it means that process is online.** It works for process or cluster. That realized in God.js realizeQueue function (for start command) and in ActionMethods.js (for restart command). **Clustered process have status online, when every cluster send signal 'ready' or 'pm2:started'**


Start/restart/stop priority behavior in pm2:

1. Start/restart priority works in case, when at least one process has 'priority' property in ecosystem

2. Stop priority works in case, when at least one process has 'stop_priority' or 'priority' property

3. If two or more processes has same priority, they starting in the same time by async

4. In case, when at least on process has not valid priority (number more than count of processes or hasn't priority) processes starting/stoping by next order: 1) processes has valid priority, 2) processes has not valid priority, 3) processes without priority
For example, Process_1.priority: 35, Process_2.priority: null, Process_3.priority: 1; Starting prio: Process_3 --------> Process_1 -------> Process_2

5. Process and cluster is online, when pm2 got 'pm2:started' or 'ready' message from this process

6. If processes started by one priority case and after that priority has been changed, restart command will restart processes by new priority order

**Comment by msupra** [comment]
Hi

Have this been implemented yet, as I have a use case for it?

Regards
Morne

**Comment by bert-huang** [comment]
Hi,

Is there any plan on merging in this pull request?

Thanks
Bert

File: lib/God.js [hunk 1]
```diff
@@ -434,10 +434,33 @@ God.prepare = function prepare (env, cb) {
         God.notify('start', clu, true);
         // here call next wihtout an array because
         // async.times aggregate the result into an array
-        return next(null, Utility.clone(clu));
+
+        realizeQueue(clu, next, false);
```
Comment (bert-huang) [inline lib/God.js:438]: Is this bit missing the `return`?

**Pull request closed without merging**